src/goto-programs/string_refine_preprocess.cpp

```diff
@@ -0,0 +1,1347 @@
+
+Function: string_refine_preprocesst::make_cprover_string_assign
+
+  Inputs: a goto_program, a position in this program, an expression and a
+          location
+
+ Outputs: an expression
+
+ Purpose: Introduce a temporary variable for cprover strings;
+          returns the cprover_string corresponding to rhs if it is a string
+          pointer and the original rhs otherwise.
+
+\*******************************************************************/
+
+exprt string_refine_preprocesst::make_cprover_string_assign(
+  goto_programt & goto_program,
+  goto_programt::instructionst::iterator & i_it,
+  const exprt & rhs,
+  source_locationt location)
+{
+  if(is_java_string_pointer_type(rhs.type()))
+  {
+    auto it=java_to_cprover_strings.find(rhs);
+    if(it!=java_to_cprover_strings.end())
+      return it->second;
+    else
+    {
+      debug() << "string_refine_preprocesst::make_cprover_string_assign "
+              << rhs.id() << eom;
+      exprt clsid=symbol_exprt("java::java.lang.String",string_typet());
+      // get_class_identifier_field(rhs, symbol_typet("java::java.lang.Object"), ns);
+      string_exprt new_rhs(rhs, clsid, true);
+      symbol_exprt lhs=new_symbol("cprover_string", new_rhs.type());
+      symbol_exprt array_lhs = new_symbol(
+        "cprover_string_array", new_rhs.content().type());
+      exprt content=new_rhs.content();
+      // This does not look nice:
+      new_rhs.op1()=array_lhs;
+      code_assignt array_assignment(array_lhs, content);
+      code_typet deref_type;
+      deref_type.return_type()=array_lhs.type();
+      code_assignt assignment(lhs, new_rhs);
+      java_to_cprover_strings[rhs] = lhs;
+
+      lhs.add_source_location() = location;
+      assignment.add_source_location() = location;
+      array_assignment.add_source_location() = location;
+      i_it->make_assignment();
+      i_it->code=array_assignment;
+      i_it=goto_program.insert_after(i_it);
+      i_it->make_assignment();
+      i_it->code=assignment;
+      i_it=goto_program.insert_after(i_it);
+      return lhs;
+    }
+  }
+  else return rhs;
+}
+
+/*******************************************************************\
+
+Function: string_refine_preprocesst::make_normal_assign
+
+  Inputs: a goto_program, a position in this program, an expression lhs,
+          a function type, a function name, a vector of arguments, a location
+          and a signature
+
+ Purpose: replace the current instruction by:
+          > lhs = function_name(arguments) : return_type @ location
+          If given, signature can force String conversion of given arguments.
+          The convention for signature is one character by argument
+          and 'S' denotes string.
+
+\*******************************************************************/
+
+void string_refine_preprocesst::make_normal_assign(
+  goto_programt & goto_program,
+  goto_programt::instructionst::iterator & i_it,
+  exprt lhs,
```

Comment:
Done in https://github.com/diffblue/cbmc/pull/378/commits/6f8db489d7c1c721461851415ef98ccda37bb3ec